criu/tls.c

```diff
@@ -0,0 +1,336 @@
+ * This function is used for cases when we would otherwise use splice()
+ * to transfer data from a TCP socket to a PIPE.
+ */
+ssize_t tls_recv_data_to_fd(int fd, unsigned long len)
+{
+	ssize_t received;
+	ssize_t total_written = 0;
+	char *buffer = xmalloc(len);
+
+	if (!buffer)
+		return -1;
+
+	while (len > total_written) {
+		received = tls_recv((buffer + total_written), (len - total_written));
+		if (received < 0) {
+			total_written = -1;
+			goto err;
+		}
+
+		int written = 0;
+		while (written < received) {
+			int ret = write(fd,
+				(buffer + total_written + written),
+				(received - written));
+
+			if (ret < 0) {
+				pr_perror("Failed writing to fd");
+				total_written = -1;
+				goto err;
+			}
+
+			written += ret;
+		}
+
+		total_written += received;
+	}
+err:
+	xfree(buffer);
+	return total_written;
+}
+
+static int tls_set_priority(gnutls_session_t *session)
+{
+	int ret;
+
+	ret = gnutls_priority_init(&priority_cache, TLS_PRIORITY, NULL);
+	if (ret == GNUTLS_E_SUCCESS)
+		ret = gnutls_priority_set(*session, priority_cache);
+
+	return ret;
+}
+
+static ssize_t _tls_data_push_cb(gnutls_transport_ptr_t ptr, const void* data, size_t len)
+{
+	int sockfd = *(int *)(ptr);
+	return send(sockfd, data, len, 0);
+}
+
+static ssize_t _tls_data_pull_cb(gnutls_transport_ptr_t ptr, void* data, size_t maxlen)
+{
+	int sockfd = *(int *)(ptr);
+	return recv(sockfd, data, maxlen, 0);
+}
+
+static int tls_handshake()
+{
+	int ret;
+
+	ret = tls_set_priority(&session);
+	if (ret != GNUTLS_E_SUCCESS) {
+		tls_perror("Failed to set priority", ret);
+		return -1;
+	}
+
+	connfdPtr = xmalloc(sizeof(int));
+	*connfdPtr = tls_sk;
+	gnutls_transport_set_ptr(session, connfdPtr);
+
+	gnutls_transport_set_push_function(session, _tls_data_push_cb);
+	gnutls_transport_set_pull_function(session, _tls_data_pull_cb);
+
+	pr_info("Perform TLS handshake\n");
+	do {
+		ret = gnutls_handshake(session);
+
+		if (gnutls_error_is_fatal(ret)) {
+			tls_perror("TLS handshake has failed", ret);
+			return -1;
+		}
+
+	} while (ret != 0);
+
+	return 0;
+}
+
+static int _tls_psk_creds_cb(gnutls_session_t session, const char *username, gnutls_datum_t *key)
+{
+	key->size = strlen(opts.tls_psk_key);
+	key->data = gnutls_malloc(key->size);
+	if (key->data == NULL) {
+		return -1;
+	}
+	memcpy(key->data, opts.tls_psk_key, key->size);
+	return 0;
+}
+
+int tls_psk_client_init(int sockfd)
+{
+	int ret;
+	gnutls_datum_t key;
+
+	tls_sk = sockfd;
+
+	pr_info("Initialize TLS client session using pre-shared key\n");
+	ret = gnutls_init(&session, GNUTLS_CLIENT);
+	if (ret != GNUTLS_E_SUCCESS) {
+		tls_perror("Failed to initialize session", ret);
+		return -1;
+	}
+
+	ret = gnutls_psk_allocate_client_credentials(&c_psk_cred);
+	if (ret != 0) {
+		tls_perror("Failed to allocate credentials structure", ret);
+		return -1;
+	}
+
+	key.size = strlen(opts.tls_psk_key);
+	key.data = xmalloc(key.size);
+	memcpy(key.data, opts.tls_psk_key, key.size);
```

Comment:
Do we need malloc() & memcpy() here, or can just pass the key?
BTW, malloc() can fail.